x-pack/plugins/alerting/public/pages/maintenance_windows/components/table_actions_popover.tsx

```diff
@@ -0,0 +1,209 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback, useMemo, useState } from 'react';
+import {
+  EuiButtonIcon,
+  EuiConfirmModal,
+  EuiContextMenuItem,
+  EuiContextMenuPanel,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiPopover,
+} from '@elastic/eui';
+import * as i18n from '../translations';
+import { MaintenanceWindowStatus } from '../../../../common';
+
+interface TableActionsPopoverProps {
+  status: MaintenanceWindowStatus;
+  onEdit: () => void;
+  onCancel: () => void;
+  onArchive: (archive: boolean) => void;
+  onCancelAndArchive: () => void;
+}
+type ModalType = 'cancel' | 'cancelAndArchive' | 'archive' | 'unarchive';
+
+export const TableActionsPopover: React.FC<TableActionsPopoverProps> = React.memo(
+  ({ status, onEdit, onCancel, onArchive, onCancelAndArchive }) => {
+    const [isPopoverOpen, setIsPopoverOpen] = useState(false);
+    const [isModalVisible, setIsModalVisible] = useState(false);
+    const [modalType, setModalType] = useState<ModalType>();
+
+    const onButtonClick = useCallback(() => {
+      setIsPopoverOpen((open) => !open);
+    }, []);
+    const closePopover = useCallback(() => {
+      setIsPopoverOpen(false);
+    }, []);
+
+    const closeModal = useCallback(() => setIsModalVisible(false), []);
+    const showModal = useCallback((type: ModalType) => {
+      setModalType(type);
+      setIsModalVisible(true);
+    }, []);
+
+    const modal = useMemo(() => {
+      const modals = {
+        cancel: {
+          title: i18n.CANCEL_MODAL_TITLE,
+          subtitle: i18n.CANCEL_MODAL_SUBTITLE,
+          cancelButton: i18n.CANCEL_MODAL_BUTTON,
+          onConfirm: () => onCancel(),
+        },
+        cancelAndArchive: {
+          title: i18n.CANCEL_AND_ARCHIVE_MODAL_TITLE,
+          subtitle: i18n.CANCEL_AND_ARCHIVE_MODAL_SUBTITLE,
+          cancelButton: i18n.CANCEL_MODAL_BUTTON,
+          onConfirm: () => onCancelAndArchive(),
+        },
+        archive: {
+          title: i18n.ARCHIVE_TITLE,
+          subtitle: i18n.ARCHIVE_SUBTITLE,
+          cancelButton: i18n.CANCEL,
+          onConfirm: () => onArchive(true),
+        },
+        unarchive: {
+          title: i18n.UNARCHIVE_MODAL_TITLE,
+          subtitle: i18n.UNARCHIVE_MODAL_SUBTITLE,
+          cancelButton: i18n.CANCEL,
+          onConfirm: () => onArchive(false),
+        },
+      };
+      let m;
+      if (isModalVisible && modalType) {
+        const modalProps = modals[modalType];
+        m = (
+          <EuiConfirmModal
+            style={{ width: 600 }}
+            title={modalProps.title}
+            onCancel={closeModal}
+            onConfirm={() => {
+              closeModal();
+              modalProps.onConfirm();
+            }}
+            cancelButtonText={modalProps.cancelButton}
```

Comment:
to make this more concise I think we can collapse some of these fields:

```tsx
// line 50:

const modals = {
  cancel: {
    props: {
      confirmButtonText: i18n.CANCEL_MODAL_TITLE,
      onConfirm: () => {
        closeModal();
        onCancel();
      },
      ...
    },
    subtitle: i18n.CANCEL_MODAL_SUBTITLE,
  },
  cancelAndArchive: {...},
  ...
};

<EuiConfirmModal
  {...modalProps.props}
  {...otherProps}
>
  <p>{modalProps.subtitle}</p>
</EuiConfirmModal>
```

Reply:
Resolved in this commit https://github.com/elastic/kibana/pull/155201/commits/9d6e6730e18bd54c5bece8e97a0c801d232edd3c